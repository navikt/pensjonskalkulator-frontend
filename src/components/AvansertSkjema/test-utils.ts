export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
