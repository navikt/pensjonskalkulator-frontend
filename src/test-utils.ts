import renderer from 'react-test-renderer'

import { render } from '@testing-library/react'

export function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => children,
    ...options,
  })

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
