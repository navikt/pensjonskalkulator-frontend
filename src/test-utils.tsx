import React, { PropsWithChildren } from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

import { PreloadedState, createListenerMiddleware } from '@reduxjs/toolkit'
import { render, RenderOptions } from '@testing-library/react'

import { getTranslation_test } from '@/utils/__tests__/test-translations'

import { createUttaksalderListener } from './state/listeners/uttaksalderListener'
import {
  setupStore,
  RootState,
  AppStore,
  AppStartListening,
} from './state/store'
import { getTranslation_nb } from './translations/nb'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  hasRouter?: boolean
}

export const swallowErrors = (testFn: () => void) => {
  const error = console.error
  console.error = () => {}
  testFn()
  console.error = error
}

export const swallowErrorsAsync = async (testFn: () => Promise<void>) => {
  const cache = console.error
  console.error = () => {}
  await testFn()
  console.error = cache
}

function generateMockedTranslations() {
  const nbTranslations: Record<string, string> = getTranslation_nb()
  const testTranslations: Record<string, string> = getTranslation_test()
  const translationsInput = { ...nbTranslations, ...testTranslations }
  const translations: Record<string, string> = {}

  for (const key in translationsInput) {
    if (
      /<(?=.*? .*?\/ ?>|br|hr|input|!--|wbr)[a-z]+.*?>|<([a-z]+).*?<\/\1>|{[^}]+}/i.test(
        translationsInput[key]
      )
    ) {
      // for html or chunks keys: results in 'my_key' : 'my_key with some <html> or {chunk}'
      translations[key] = translationsInput[key]
    } else {
      // results in 'my_key' : 'my_key'
      translations[key] = key
    }
  }
  return translations
}

// Return an object with the store and all of RTL's query functions
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState, true),
    hasRouter = true,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <IntlProvider locale="nb" messages={generateMockedTranslations()}>
          {hasRouter ? <MemoryRouter>{children}</MemoryRouter> : children}
        </IntlProvider>
      </Provider>
    )
  }

  const listenerMiddleware = createListenerMiddleware()
  createUttaksalderListener(
    listenerMiddleware.startListening as AppStartListening
  )

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'

export { renderWithProviders as render }
