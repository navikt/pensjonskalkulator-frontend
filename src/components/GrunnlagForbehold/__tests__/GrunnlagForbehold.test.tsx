import { GrunnlagForbehold } from '..'
import { render, screen } from '@/test-utils'

describe('GrunnlagForbehold', () => {
  it('rendrer riktig', () => {
    const { asFragment } = render(<GrunnlagForbehold headingLevel="2" />)
    expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(
      'grunnlag.forbehold.title'
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
