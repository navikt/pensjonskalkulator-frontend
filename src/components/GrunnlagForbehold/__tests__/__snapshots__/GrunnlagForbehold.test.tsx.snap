// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GrunnlagForbehold > rendrer riktig 1`] = `
<DocumentFragment>
  <section
    class="_section_1598cd"
  >
    <h2
      class="_heading_1598cd navds-heading navds-heading--medium"
    >
      grunnlag.forbehold.title
    </h2>
    <p
      class="_text_1598cd navds-body-long navds-body-long--medium"
    >
      grunnlag.forbehold.ingress_1
      <a
        class="navds-link navds-link--action navds-link--inline-text"
        href="/forbehold"
        target="_blank"
      >
        grunnlag.forbehold.link
        <svg
          aria-labelledby="title-r0"
          fill="none"
          focusable="false"
          height="1.25rem"
          role="img"
          viewBox="0 0 24 24"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="title-r0"
          >
            application.global.external_link
          </title>
          <path
            clip-rule="evenodd"
            d="M20.532 3.471A.748.748 0 0 1 20.75 4v7.5a.75.75 0 0 1-1.5 0V5.81l-8.72 8.72a.75.75 0 1 1-1.06-1.06l8.72-8.72H12.5a.75.75 0 0 1 0-1.5H20c.206 0 .393.083.529.218l.001.002.002.001ZM4.75 9A.25.25 0 0 1 5 8.75h7a.75.75 0 0 0 0-1.5H5A1.75 1.75 0 0 0 3.25 9v10c0 .966.784 1.75 1.75 1.75h10A1.75 1.75 0 0 0 16.75 19v-7a.75.75 0 0 0-1.5 0v7a.25.25 0 0 1-.25.25H5a.25.25 0 0 1-.25-.25V9Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </a>
      <br />
      <br />
      Denne kalkulatoren er under utvikling og har begrenset funksjonalitet. Du kan gj√∏re mer spesifisert beregning i 
      <a
        class="navds-link navds-link--action navds-link--inline-text"
        href="https://www.nav.no/pselv/simulering.jsf"
        target="_blank"
      >
        detaljert kalkulator
        <svg
          aria-labelledby="title-r1"
          fill="none"
          focusable="false"
          height="1.25rem"
          role="img"
          viewBox="0 0 24 24"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="title-r1"
          >
            application.global.external_link
          </title>
          <path
            clip-rule="evenodd"
            d="M20.532 3.471A.748.748 0 0 1 20.75 4v7.5a.75.75 0 0 1-1.5 0V5.81l-8.72 8.72a.75.75 0 1 1-1.06-1.06l8.72-8.72H12.5a.75.75 0 0 1 0-1.5H20c.206 0 .393.083.529.218l.001.002.002.001ZM4.75 9A.25.25 0 0 1 5 8.75h7a.75.75 0 0 0 0-1.5H5A1.75 1.75 0 0 0 3.25 9v10c0 .966.784 1.75 1.75 1.75h10A1.75 1.75 0 0 0 16.75 19v-7a.75.75 0 0 0-1.5 0v7a.25.25 0 0 1-.25.25H5a.25.25 0 0 1-.25-.25V9Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </a>
    </p>
  </section>
</DocumentFragment>
`;
